### PR TITLE
Feat/edit product

### LINKED ISSUE #123: [FEAT] Edit User Products with Update Timestamp
# API Endpoint: Edit User Product

## Description

Develop an API endpoint to edit user product_name, price, description, and tag and create a timestamp for updates.

## Acceptance Criteria

- The endpoint should be accessible at `api/v1/products/{id}`
- The endpoint should accept HTTP PUT requests.
- The endpoint should be secured to ensure that only authenticated users can edit products.
- Proper authentication mechanisms (e.g., JWT, OAuth2) should be implemented.
- Requests to the endpoint must include a valid authentication token in the Authorization header.
  ```
  Authorization: Bearer <token>
  ```

### Request Body:
```json
{
  "product_name": "Updated Product Name",
  "price": 600.0,
  "description": "Updated Product Description",
  "tag": "New Tag"
}
```

### Response:
On a successful update, the API should return a 200 OK status code. The response body should contain the updated product details.
```json
{
  "id": 1,
  "product_name": "Updated Product Name",
  "price": 600.0,
  "description": "Updated Product Description",
  "tag": "New Tag",
  "created_date": "2024-07-17T12:34:56Z",
  "updated_date": "2024-07-18T14:20:00Z"
}
```

### Validation:
If any required field is missing or invalid, the API should return a 400 Bad Request status code with an appropriate validation error message.
```json
{
  "status_code": 400,
  "message": ["product_name is required", "price must be a positive number"],
  "error": "Bad Request"
}
```

## Purpose

Provides a backend service that allows users to edit their products, including updating the name, price, description, and tag. This is useful because users can have multiple products, and product details can change over time. The endpoint also enables users to see the specific time when a product was last updated.

## Requirements

1. **Endpoint**: The API should expose an endpoint at `api/v1/products/{id}`.
2. **Request Method**: The endpoint should accept HTTP PUT requests.
3. **Authentication**: Implement proper authentication mechanisms (e.g., JWT, OAuth2) to secure the endpoint.
4. **Authorization Header**: Requests to the endpoint must include a valid authentication token in the Authorization header.
5. **Request Body**: Accept a JSON body with the fields `product_name`, `price`, `description`, and `tag`.

## Expected Outcome

The API should update the product details based on the provided information. The response should include updated product details such as `id`, `product_name`, `price`, `description`, `tag`, `created_date`, and `updated_date`.

### Authentication Testing:
- Test the endpoint with a valid authentication token to ensure access is granted.
- Test the endpoint without an authentication token or with an invalid token to ensure access is denied.

### Parameter Validation:
- Test the endpoint with valid fields to ensure it updates the product correctly.
- Test the endpoint with missing or invalid fields to ensure it returns a 400 Bad Request status code with an appropriate error message.

### Response Verification:
- Test the endpoint with valid input values and verify that the response includes the updated product details.
- Test the endpoint with an id that does not exist and verify that the response is appropriate (e.g., a 404 Not Found status code).

### Edge Cases:
- Test the endpoint with special characters and boundary values in the `product_name`, `description`, and `tag` fields to ensure the update functionality is robust.

### PR DESCRIPTION
**Pull request by urizennnn**
## Description
This PR implements a new endpoint PUT /api/v1/products/{product_id} that allows updating existing product details. The endpoint provides the following functionality:

Updates product information based on the provided product ID
Validates input data to ensure data integrity
Handles authentication and authorization to ensure only authorized users can update products
Returns appropriate responses for successful updates, validation errors, and not found scenarios

Key features implemented:

Input validation for product name, description, and price
Authorization check to ensure only the product owner can update it
Database update operation with error handling
Structured JSON responses for various scenarios
​
## Related Issue (Link to Github issue)
closes #123 #120 
​
## How Has This Been Tested?
It has been tested using postman

### **​Success Response**
image

### **Product not found response**
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [] Breaking change 
​
## Checklist:

- [X] My code follows the code style of this project.
- [X] I have updated the documentation accordingly.
- [] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: tests/test_product/product_test.go [hunk 1]
```diff
@@ -132,3 +132,123 @@ func TestProductCreate(t *testing.T) {
...
+// 		Name:        "Nike SB",
+// 		Description: "One of the best, common and cloned nike product of all time",
+// 		Price:       190.33,
+// 	}
+//
+// 	product := product.Controller{Db: db, Validator: validatorRef, Logger: logger}
+// 	productUrl := r.Group(fmt.Sprintf("%v", "/api/v1"), middleware.Authorize(db.Postgresql))
+// 	{
+// 		productUrl.POST("/products", product.CreateProduct)
+// 	}
```
Comment (Cyberguru1) [inline tests/test_product/product_test.go]: remove this comments if you don't need them again

**Review by Cyberguru1 — CHANGES_REQUESTED** [review]

**Comment by Cyberguru1** [comment]
Please add more description to this PR

**Comment by urizennnn** [comment]
> Please add more description to this PR

done i will close the other one since the entire changes are present here too

**Review by Cyberguru1 — APPROVED** [review]
lgtm
